qutip/nonmarkov/bofin.py

```diff
@@ -0,0 +1,1445 @@
+            "zvode",
+            method=self.options.method,
+            order=self.options.order,
+            atol=self.options.atol,
+            rtol=self.options.rtol,
+            nsteps=self.options.nsteps,
+            first_step=self.options.first_step,
+            min_step=self.options.min_step,
+            max_step=self.options.max_step,
+        )
+
+        self._ode = solver
+        self.RHSmat = RHSmat
+
+    def extract_ado(self, ado_state, idx):
+        """
+        Extract a Qobj representing specified ADO from a full representation of
+        the ADO states, as returned by :meth:`.steady_state` or :meth:`.run`.
+
+        Parameters
+        ----------
+        ado_state : numpy array
+            A full representation of the ADO state.
+        idx : int
+            The index of the ADO to extract.
+
+        Returns
+        -------
+        Qobj
+            A :obj:`Qobj` representing the state of the specified ADO.
+        """
+        return Qobj(ado_state[idx, :].T, dims=self._sys_dims)
+
+    def steady_state(
+        self,
+        use_mkl=False, mkl_max_iter_refine=100, mkl_weighted_matching=False
+    ):
+        """
+        Compute the steady state of the system.
+
+        Parameters
+        ----------
+        use_mkl : bool, default=False
+            Whether to use mkl or not. If mkl is not installed or if
+            this is false, use the scipy splu solver instead.
+
+        mkl_max_iter_refine : Int
+            Parameter for the mkl LU solver. If pardiso errors are returned
+            this should be increased.
+
+        mkl_weighted_matching : Boolean
+            Setting this true may increase run time, but reduce stability
+            (pardisio may not converge).
+
+        Returns
+        -------
+        steady_state : Qobj
+            The steady state density matrix of the system.
+
+        full_ado_state : Numpy array
+            Array of the the steady-state and all ADOs.
+            The state of a particular ADO may be extracted from the
+            full state by calling :meth:`.extract_ado`.
+        """
+        n = self._sys_shape
+
+        b_mat = np.zeros(n ** 2 * self._n_ados, dtype=complex)
+        b_mat[0] = 1.0
+
+        L = deepcopy(self.RHSmat)
+        L = L.tolil()
+        L[0, 0: n ** 2 * self._n_ados] = 0.0
+        L = L.tocsr()
+        L += sp.csr_matrix((
+            np.ones(n),
+            (np.zeros(n), [num * (n + 1) for num in range(n)])
+        ), shape=(n ** 2 * self._n_ados, n ** 2 * self._n_ados))
+
+        if mkl_spsolve is not None and use_mkl:
+            L.sort_indices()
+            solution = mkl_spsolve(
+                L,
+                b_mat,
+                perm=None,
+                verbose=True,
+                max_iter_refine=mkl_max_iter_refine,
+                scaling_vectors=True,
+                weighted_matching=mkl_weighted_matching,
+            )
+        else:
+            L = L.tocsc()
+            LU = splu(L)
```

Comment:
Very minor comment: Might be worth adding support for lgmres and other methods at some point.